src/storage/src/object/s3.rs

```diff
@@ -26,80 +29,100 @@ pub struct S3ObjectStore {
+                .bucket(&self.bucket)
+                .body(SdkBody::from(obj).into())
+                .key(path)
+                .send()
+                .await?;
+            Ok(())
+        }
     }
 
     /// Amazon S3 doesn't support retrieving multiple ranges of data per GET request.
-    async fn read(&self, path: &str, block_loc: Option<BlockLocation>) -> ObjectResult<Bytes> {
-        let req = self.client.get_object().bucket(&self.bucket).key(path);
-
-        let range = match block_loc.as_ref() {
-            None => None,
-            Some(block_location) => block_location.byte_range_specifier(),
-        };
-
-        let req = if let Some(range) = range {
-            req.range(range)
-        } else {
-            req
-        };
-
-        let resp = req.send().await?;
-        let val = resp.body.collect().await?.into_bytes();
-
-        if block_loc.is_some() && block_loc.as_ref().unwrap().size != val.len() {
-            return Err(ObjectError::internal(format!(
-                "mismatched size: expected {}, found {} when reading {} at {:?}",
-                block_loc.as_ref().unwrap().size,
-                val.len(),
-                path,
-                block_loc.as_ref().unwrap()
-            )));
+    fn read<'a>(&'a self, path: &'a str, block_loc: Option<BlockLocation>) -> Self::BytesFuture<'a> {
+        async move {
+            let req = self.client.get_object().bucket(&self.bucket).key(path);
+
+            let range = match block_loc.as_ref() {
+                None => None,
+                Some(block_location) => block_location.byte_range_specifier(),
+            };
+
+            let req = if let Some(range) = range {
+                req.range(range)
+            } else {
+                req
+            };
+
+            let resp = req.send().await?;
+
+            let val = resp.body.collect().await?.into_bytes();
+
+            if block_loc.is_some() && block_loc.as_ref().unwrap().size != val.len() {
+                return Err(ObjectError::internal(format!(
+                    "mismatched size: expected {}, found {} when reading {} at {:?}",
+                    block_loc.as_ref().unwrap().size,
+                    val.len(),
+                    path,
+                    block_loc.as_ref().unwrap()
+                )));
+            }
+            Ok(val)
         }
-        Ok(val)
     }
 
-    async fn readv(&self, path: &str, block_locs: Vec<BlockLocation>) -> ObjectResult<Vec<Bytes>> {
-        let futures = block_locs
-            .into_iter()
-            .map(|block_loc| self.read(path, Some(block_loc)))
-            .collect_vec();
-        try_join_all(futures).await
+    fn readv<'a>(&'a self, path: &'a str, block_locs: Vec<BlockLocation>) -> Self::BytesVecFuture<'a> {
+        async move {
+            let futures = block_locs
+                .into_iter()
+                .map(|block_loc| self.read(path, Some(block_loc)))
+                .collect_vec();
```

Comment:
This is the failed source: the `self.read()` seems like can not capture the upper lifetime. expend `self.read()` can solve this, but `self.read()` is non-trival in s3 impl.